overlord/snapstate/check_snap.go

```diff
@@ -451,32 +576,81 @@ var (
 	findGid = osutil.FindGid
 )
 
-// TODO: keep this unsupported until it is complete!
-var systemUsernamesSupported = false
-
 func checkSystemUsernames(si *snap.Info) error {
-	// TODO: keep this unsupported until it is complete!
-	if !systemUsernamesSupported && len(si.SystemUsernames) > 0 {
-		return fmt.Errorf("system usernames are not yet supported")
+	// No need to check support if no system-usernames
+	if len(si.SystemUsernames) == 0 {
+		return nil
+	}
+
+	// TODO: wrap this invocation differently in a sandbox helper
+	// Run /.../snap-seccomp version-info
+	path, err := cmd.InternalToolPath("snapd")
+	if err != nil {
+		return err
+	}
+
+	vi, err := interfaces.SeccompCompilerVersionInfo(filepath.Join(filepath.Dir(path), "snap-seccomp"))
+	if err != nil {
+		return fmt.Errorf("cannot obtain seccomp compiler information: %v", err)
+	}
+
+	// If the system doesn't support robust argument filtering then we
+	// can't support system-usernames
+	err = seccomp_compiler.VersionInfo(vi).SupportsRobustArgumentFiltering()
+	if err != nil {
+		if re, ok := err.(*seccomp_compiler.BuildTimeRequirementError); ok {
+			return fmt.Errorf("snap %q system usernames require a snapd built against %s", si.InstanceName(), re.RequirementsString())
+		}
+		return err
 	}
 
 	for _, user := range si.SystemUsernames {
-		if !supportedSystemUsernames[user.Name] {
+		var id uint32
+		id, ok := supportedSystemUsernames[user.Name]
+		if !ok {
 			return fmt.Errorf(`snap %q requires unsupported system username "%s"`, si.InstanceName(), user.Name)
 		}
 
 		switch user.Scope {
 		case "shared":
 			_, uidErr := findUid(user.Name)
 			_, gidErr := findGid(user.Name)
-			if uidErr != nil || gidErr != nil {
+			if uidErr != nil && gidErr != nil {
+				// If neither exist, create the user and group
+				extrausers := !release.OnClassic
+				err := osutil.UserGroupAdd(user.Name, id, extrausers)
+				if err != nil {
+					return err
+				}
+			} else if uidErr != nil || gidErr != nil {
 				return fmt.Errorf(`snap %q requires that both the "%s" system user and group are present on the system.`, si.InstanceName(), user.Name)
```

Comment:
it seems that if we failed once to create both we will be stuck forever on this error? shouldn't we try instead to create the missing one only?

Reply:
Also shouldn't we follow the style of other things and have a idempotent osutil.EnsureUserGroup and make it the problem of that function to do the right thing?

Reply:
that would also avoid the find* here,  also I think we should distinguish the various user.Unknown* errors from other the system is broken is errors.

Reply:
IMHO, no. The idea is that we have both already or neither (where we will create it) and the only time just one isn't there is if the administrator removed it explicitly. IMO we shouldn't assume we know better than the administrator. It does mean that if useradd fails in UserGroupAdd, then the group is left lying around. We could fix UserGroupAdd() to use groupdel, but alas it doesn't support --extrausers.

Reply:
I see, my point about moving this to EnsureUseGroup stands. We are trying to add remove-users support @mvo5 knows more about what was added to support that.

Reply:
> I see, my point about moving this to EnsureUseGroup stands. We are trying to add remove-users support @mvo5 knows more about what was added to support that.

I can adjust to have:
```
case "shared":
        err := osutil.EnsureUserGroup(user.Name, id, extrausers)
        if err != nil {
                return err
        }
case "private", "external":
...
```

Reply:
Done